WIP: Ingress node firewall APIs review

msherif1234 opened the pull request:
request to review new operator APIs, implementation is based on operator-sdk framework
repo where the APIs and controller are implemented is 
https://github.com/openshift/ingress-node-firewall

review comments will be addressed in PR against ingress-node-firewall repo directly I shared in api forum for reviews
Signed-off-by: Mohamed Mahmoud <mmahmoud@redhat.com>

JoelSpeed commented on ingress-node-firewall/v1alpha1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "ingress-nodefw.ingress-nodefw", Version: "v1alpha1"}
```
Should the group not be an FQDN with a valid TLD?

msherif1234 replied:
not sure, I checked some of the operators we ahve in OCP and it seems free form , can u show me an example of what u want to be done ?

msherif1234 replied:
`ingress.node.firewall.openshift.io` is good ?

martinkennelly replied:
I think `ingressnodefirewall.openshift.io`. Using dots to separate "ingress node firewall" may erroneously tell the user there are multiple subdomains instead of one.

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_webhook.go:
```diff
@@ -0,0 +1,256 @@
+var _ webhook.Validator = &IngressNodeFirewall{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *IngressNodeFirewall) ValidateCreate() error {
```
You should avoid having functions associated with your API package. It brings in additional dependencies and makes it difficult for people to import and re-use your API types.

You can wrap the webhook in an additional struct to avoid this pattern

msherif1234 replied:
can u pls expand on this more ?

martinkennelly replied:
Ill do this. Thanks Joel.

martinkennelly replied:
PR: https://github.com/openshift/ingress-node-firewall/pull/81

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+
+// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
+type IngressNodeFirewallICMPRule struct {
+	// ICMPType define ICMP Type Numbers (RFC 792).
```
I would add some additional context here about examples of what possible values might be and what they would mean to an end user

Are you sure you want to allow all of the ICMP types? Is there a way to make this API nicer by allowing a subset and using human readable names rather than numbers?

msherif1234 replied:
user can use any icmpv4 type and/or code based on https://www.rfc-editor.org/rfc/rfc792.html
and icmpv6 types and/or codes bases on https://www.rfc-editor.org/rfc/rfc4443.html

JoelSpeed replied:
Reading that list, it would seem there is a list of valid values of ICMPv4
```
Summary of Message Types

    0  Echo Reply

    3  Destination Unreachable

    4  Source Quench

    5  Redirect

    8  Echo

   11  Time Exceeded

   12  Parameter Problem

   13  Timestamp

   14  Timestamp Reply

   15  Information Request

   16  Information Reply
```
We can make an enum out of the numeric values, or could make it easier to understand, out of the textual values.

msherif1234 replied:
we have types and codes we will endup with huge enum list one for v4 and another for v6 we already allow user entrying number for dstports and cidr so I wil ljust go with min and make sure its fits in 255 byte size

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
+type IngressNodeFirewallICMPRule struct {
+	// ICMPType define ICMP Type Numbers (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=255
+	// +kubebuilder:validation:Minimum:=0
+	// +optional
```
You have optional and required, which one is it?

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+	// +optional
+	ICMPType uint8 `json:"icmpType,omitempty"`
+
+	// ICMPCode define ICMP Code ID (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=16
+	// +kubebuilder:validation:Minimum:=0
+	// +optional
```
Same comments as above, hard to know what this means so perhaps you can make this easier by naming them or providing aliases at least

Also, optional and required?

msherif1234 replied:
can u explain what u meant by aliases ? it will be a huge list of strings to cover both icmpv4 and icmpv6

JoelSpeed replied:
Yeah I'm suggesting an enum, if possible, the types have a defined set within the spec, do the codes as well?

JoelSpeed replied:
At a minimum, the code and type both seem to have a limit. They are 8 bytes of data at most so, if we keep this numeric, the value must be 0-255 right?

msherif1234 replied:
I do already have limit for the Type and the code

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+	//Ports can be a single port or range of ports using start-end range format
+	//+kubebuilder:validation:Optional
+	// +optional
+	Ports string `json:"ports,omitempty"`
```
You might want to use an IntOrString type for this

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+
+// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols
+type IngressNodeFirewallProtoRule struct {
+	//Ports can be a single port or range of ports using start-end range format
```
You should extend this description to explain why you would want to use a single port or a range

msherif1234 replied:
just to give user more flexibility

JoelSpeed replied:
We like to explain the use case for different options within the API so that the generated docs can help a user understand why they would make the choice between the options. Eg if you want the rule to apply to a single port, set single port, if you need to cover a range of ports, you can set a start-end range.

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols
+type IngressNodeFirewallProtoRule struct {
+	//Ports can be a single port or range of ports using start-end range format
+	//+kubebuilder:validation:Optional
+	// +optional
```
You only need one optional tag, use `+optional`

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+
+// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
+type IngressNodeFirewallProtocolRule struct {
+	// Order define order of execution of ingress firewall rules .
```
What happens if multiple rules specify the same order number?

msherif1234 replied:
validation webhook detect and error the config

JoelSpeed replied:
If you made this a `+listType=map`, `+listMapKey=order` then the user wouldn't be able to do this and the OpenAPI schema would reject it, you then wouldn't need a webhook I believe

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+	// +kubebuilder:validation:Minimum:=1
+	Order uint32 `json:"order"`
+
+	// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols.
```
What happens when this isn't provided?

msherif1234 replied:
based on the protocol it will be wildcard dstport or code/type and action will apply with those fields don'tcare

JoelSpeed replied:
This should be explained in the godoc, `when omitted, ...`

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+	// +optional
+	ProtocolRule *IngressNodeFirewallProtoRule `json:"protoRule,omitempty"`
+
+	// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols.
```
What happens when this isn't specified?

Can this be specified at the same time as ProtocolRule?

msherif1234 replied:
I need your suggestion I was asked to try to make the config protocol specific but I ugly this ugly so what is ur thoughts ?

JoelSpeed replied:
If you want to make it based on the protocol, you need to create a discriminated union, https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions

Then you have a field for each protocol type and the specific information for that protocol within it

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+
+	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="icmp";"icmpv6";"tcp";"udp";"sctp"
```
Enums should be PascalCase. You've described that in the comment but the enum values are wrong

msherif1234 replied:
so ICMP or Icmp ?

martinkennelly replied:
imo, should be ICMP

JoelSpeed replied:
This is the expected format for PascalCase with acronyms
```
	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Enum="ICMP";"ICMPv6";"TCP";"UDP";"SCTP"
```

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+
+	// Action can be allow or deny, default action is deny
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="allow";"deny"
```
Enum must be pascal case

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+	// Action can be allow or deny, default action is deny
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="allow";"deny"
+	// +optional
```
Optional and required again

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+)
+
+// IngressNodeFirewallActionType indicates whether an IngressNodeFirewallRule allows or denies traffic
+// +kubebuilder:validation:Pattern=`^allow|deny$`
```
You should use an enum rule here instead, also, if you specify it here, you don't need it where this is used on line 70

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+	// +kubebuilder:validation:one-of[]=Type=string;Format:cidr
+	SourceCIDRs []string `json:"sourceCIDRs"`
+	// FirewallProtocolRules is A list of per protocol ingress node firewall rules
+	FirewallProtocolRules []IngressNodeFirewallProtocolRule `json:"rules,omitempty"`
```
Must the list have one entry per protocol? In which case this will want to be a map type list (`listType=map`) with a list map key (`listMapKey=protocol`)

msherif1234 replied:
no u can have multiple tcp entries for example
```yaml
 - order: 20
      protocol: tcp
      protoRule:
        ports: 800
      action: deny
 - order: 25
     protocol: tcp
      protoRule:
        ports: 900
      action: allow
```

JoelSpeed replied:
The order must be unique though? In which case the map key is order

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+type IngressNodeFirewallSpec struct {
+	// Selects node(s) where ingress firewall rules will be applied to
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```
Maybe you want to use a `metav1.LabelSelector` type to match other patterns

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+
+	// A list of ingress firewall policy rules.
+	// empty list indicates no ingress firewall i.e allow all incoming traffic.
+	// +kubebuilder:validation:Optional
```
Remove this, you already have `+optional`

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,163 @@
+	// empty list indicates the firewall policy applied on all interfaces
+	// +kubebuilder:validation:Optional
+	// +optional
+	// +nullable
```
Does this need to be nullable?

Why is this a pointer? What are you trying to distinguish between an empty list and a nil list?

msherif1234 replied:
we won't allow nil value to apply to all interfaces otherwise we apply to what is the list

msherif1234 replied:
we changed this to list of minItem of 1 and make it required to simplify the design

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewallconfig_types.go:
```diff
@@ -0,0 +1,64 @@
+type IngressNodeFirewallConfigSpec struct {
+	// Nodes where the daemonset will be running on.
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```
Maybe a metav1.LabelSelector

msherif1234 replied:
Ikept the nodeSelector in firewallconfig types as map[string]string to match up with DaemonSet Spec.Template.Spec.NodeSelector

JoelSpeed replied:
You should add a note to the godoc for developers to help them understand that motivation. Any comment prefixed with `+` won't end up in the generated docs

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewallnodestate_types.go:
```diff
@@ -0,0 +1,73 @@
+
+	// A list of ingress firewall policy rules.
+	// empty list indicates no ingress firewall i.e allow all incoming traffic.
+	// +kubebuilder:validation:Optional
```
You already have `+optional`

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewallnodestate_types.go:
```diff
@@ -0,0 +1,73 @@
+
+	// A list of interfaces where the ingress firewall policy will be applied on.
+	// empty list indicates the firewall policy applied on all interfaces
+	// +kubebuilder:validation:Optional
+	// +optional
+	// +nullable
+	Interfaces *[]string `json:"interfaces"`
```
Same questions as above, why nullable, why pointer?

msherif1234 commented:
I addressed some of your comments here https://github.com/openshift/ingress-node-firewall/pull/76 maybe can continue there as it easier to verify and test the changes ?

msherif1234 commented:
I kept the nodeSelector in firewallconfig types as `map[string]string` to match up with DaemonSet `Spec.Template.Spec.NodeSelector`

JoelSpeed commented:
Can you update this PR with the changes as described, lets keep the conversation here until we are happy with the design, then we can move across to the other repo

Keeping it here allows me to see the entire API, not just the diffed sections in isolation

msherif1234 commented:
> Can you update this PR with the changes as described, lets keep the conversation here until we are happy with the design, then we can move across to the other repo
> 
> Keeping it here allows me to see the entire API, not just the diffed sections in isolation

Thanks, I uploaded the new changes webhook change will be address in separate PR by @martinkennelly

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+
+// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
+type IngressNodeFirewallICMPRule struct {
+	// ICMPType define ICMP Type Numbers (RFC 792).
```
Perhaps we can expand the comment to help a user understand what would happen when this is set and when it is left empty

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+	// +kubebuilder:validation:Minimum:=0
+	ICMPType uint8 `json:"icmpType,omitempty"`
+
+	// ICMPCode define ICMP Code ID (RFC 792).
```
Perhaps we can expand the comment to help a user understand what would happen when this is set and when it is left empty

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+
+// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols
+type IngressNodeFirewallProtoRule struct {
+	// Ports can be a single port or range of ports using start-end range format
+	// if admin wanted to filter single port then setting a single port as an integer for example Ports: 80
+	// if admin wanted to filter range of ports then using "start-end" range string format for example Ports: "80-100"
```
Just want to tidy this up a little, for example, you should use the json tag name not the go field name, as users interact with the json names only
```suggestion
	// ports defines either a single port or a range of ports to apply a protocol rule too.
	// To filter a single port, set a single port as an integer value. For example ports: 80.
	// To filter a range of ports, use a "start-end" range, string format. For example ports: "80-100".
```

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+
+// IngressNodeProtocolConfig is a discriminated union of protocol's specific configuration.
+// +union
+type IngressNodeProtocolConfig struct {
```
You should add a discriminator to this that is an enum of the valid types, then for each type, you need to add a matching field, so 

```
config:
  type: TCP
  tcp:
    ports: 80
```

msherif1234 replied:
ok fixed that so it looks like this
```yaml
      protocolConfig:
        protocol: TCP
        protoRule:
          ports: "100-200"
```

JoelSpeed replied:
The union members must match the protocol value, you should have one entry for each discriminant assuming you need config for that discriminant
```
 protocolConfig:
        protocol: TCP
        tcp:
          ports: "100-200"
```

Looking at this, would ports better be a list of int or string?
Wondering if this would be more intuitive and would allow individual ports as well as a range 
```
ports:
  - 80
  - 8080
  - 100-200
```

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+
+// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
+type IngressNodeFirewallProtocolRule struct {
+	// Order define order of execution of ingress firewall rules .
```
You should explain in the godoc for the field that the minimum value is 1 and that all order values must be unique

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+	// +kubebuilder:validation:Minimum:=1
+	Order uint32 `json:"order"`
+
+	// ProtocolConfig is a discriminated union of protocol's specific configuration.
+	// if not specificed matching will be based on protocol value and the protocol configs like dstPort/type/code will be ignored
```
```suggestion
	// protocolConfig configures protocol specific configuration for ...
	// If not specificed, packet matching will be based on the protocol value and protocol configuration, such as dstPort/type/code, will be ignored
```

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+	// +kubebuilder:validation:Optional
+	ProtocolConfig IngressNodeProtocolConfig `json:"protocolConfig"`
+
+	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="ICMP";"ICMPv6";"TCP";"UDP";"SCTP"
+	Protocol IngressNodeFirewallRuleProtocolType `json:"protocol"`
```
Isn't this the protocol discrminator? This should be in the `IngressNodeProtocolConfig`

msherif1234 replied:
yep fixed will update the diffs with latest once I addressed the following comments

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+	// +kubebuilder:validation:Enum="ICMP";"ICMPv6";"TCP";"UDP";"SCTP"
+	Protocol IngressNodeFirewallRuleProtocolType `json:"protocol"`
+
+	// Action can be Allow or Deny, default action is deny
```
Use the json tag name and complete sentences, remember these comments form part of the user documentation
```suggestion
	// action can be Allow or Deny, default action is Deny.
```

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+	// +kubebuilder:validation:MinItems:=1
+	SourceCIDRs []string `json:"sourceCIDRs"`
+	// FirewallProtocolRules is A list of per protocol ingress node firewall rules
+	// +kubebuilder:listType:=map
+	// +kubebuilder:listMapKey:=order
```
These don't use the the kubebuilder format
```suggestion
	// +listType=map
	// +listMapKey=order
```

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go:
```diff
@@ -0,0 +1,169 @@
+// IngressNodeFirewallRules define ingress node firewall rule
+type IngressNodeFirewallRules struct {
+	// SourceCIDRS is A list of CIDRs from which we apply node firewall rule
+	// +kubebuilder:validation:one-of[]=Type=string;Format:cidr
```
Have you tested that this validation works as expected?

msherif1234 replied:
no that didn't work really we do verify the CIDR in the webhook I should remove this

msherif1234 replied:
removed that check will investigate further later for now we are protected with the verification webhook

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewall_types_methods.go:
```diff
@@ -0,0 +1,58 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func (p IngressNodeFirewallProtoRule) IsRange() bool {
```
Should these functions be removed now the webhooks are in a different package?

msherif1234 replied:
they are used with ports

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewallconfig_types.go:
```diff
@@ -0,0 +1,64 @@
+
+// IngressNodeFirewallConfigSpec defines the desired state of IngressNodeFirewallConfig
+type IngressNodeFirewallConfigSpec struct {
+	// Nodes where the daemonset will be running on.
```
This should match the json tag
```suggestion
	// nodeSelector is used to select which Nodes the <what> DaemonSet will be run on.
```

JoelSpeed commented on ingress-node-firewall/v1alpha1/ingressnodefirewallconfig_types.go:
```diff
@@ -0,0 +1,64 @@
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// tolerations is a list of tolerations applied to ingress node firewall controller
+	// deployments.
```
Deploymenys or daemonsets? Why are we mixing the two?

msherif1234 replied:
sorry will fix

msherif1234 commented on ingress-node-firewall/v1alpha1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "ingressnodefirewall.openshift.io", Version: "v1alpha1"}
```
not aware of thing that can be grouped with this feature atm if u think its might be better to use general path like `network.openshift.io` pls let me know

deads2k replied:
We currently lack a registry of chosen CR names, so having a significant group like network.openshift.io split into some things in openshift/api and some things in other repos, without any central registry is a problem for maintenance for the OCP networking team.  It's not pretty to have lots of small api groups, but it's better that conflicting schemas.

There is a PR from long ago where I floated an idea for a registry: https://github.com/openshift/api/pull/633

It could be built out, but it would take a fairly significant investment.

msherif1234 replied:
@deads2k so you are saying stay they way we are for ingress node firewall APIs group right ?

JoelSpeed replied:
Yep, lets keep it as `ingressnodefirewall` for this API
